### PR TITLE
Improve type checking fun objects

### LINKED ISSUE #108: type checking returning a fun object
I believe these examples should pass type checking, however they fail with weird errors

```
-spec nil() -> [].
nil() -> [].

-spec return_fun() -> term().
return_fun() -> fun nil/0.

-spec return_fun2() -> fun().
return_fun2() -> fun nil/0.
```
yields
```
The function nil on line 57 is expected to return INVALID-FORM:[{type,32,bounded_fun,[{type,32,'fun',[{type,32,product,[]},{type,32,nil,[]}]},[]]}]: but it returns term()
The function nil on line 60 is expected to return INVALID-FORM:[{type,32,bounded_fun,[{type,32,'fun',[{type,32,product,[]},{type,32,nil,[]}]},[]]}]: but it returns fun()
```

both functions return `fun(() -> [])` type which is a subtype of both `term()` and `fun()`.
I suspect gradualizer currently checks the opposite (via `any_subtype`), whether `term()` or `fun()` is a subtype of `fun(() -> [])`

### PR DESCRIPTION
**Pull request by gomoripeti**
Fixes #108 

More test cases are required, and I might got the subtyping wrong.

E.g. I'm not sure if this is the intended behaviour but the below certainly does not work with this change:
`fun( (integer()) -> integer(); (float()) -> float() ) <: fun( (number()) -> number() )`
because the later can be transformed as
`fun( (integer() | float()) -> number() ) == fun( (integer()) -> number(); (float()) -> number() )`

This whole fun subtyping looks similar to subtyping of maps.

### CODE REVIEW

File: src/typechecker.erl [hunk 1]
```diff
@@ -589,7 +597,11 @@ expand_builtin_aliases({type, Ann, iolist, []}) ->
              {type, Ann, iolist, []}],
     {type, Ann, maybe_improper_list, [{type, Ann, union, Union}]};
 expand_builtin_aliases({type, Ann, function, []}) ->
-    {type, Ann, 'fun', []};
+    {type, Ann, 'fun', [{type, Ann, any}, {type, Ann , any,[]}]};
+expand_builtin_aliases({type, Ann, 'fun', []}) ->
+    %% `fun()' is not a built-in alias per the erlang docs
+    %% but it is equivalent with `fun((...) -> any())'
```
[the remark this answers is not in the document]
Comment (gomoripeti) [reply src/typechecker.erl]: this step is part of normalisation only; in general we want to pretty-print the original types, so I think this is not an issue.

File: src/typechecker.erl [hunk 2]
```diff
@@ -149,6 +149,14 @@ compat_ty(Ty1, {ann_type, _, [{var, _, Var}, Ty2]}, A, TEnv) ->
 
 % TODO: There are several kinds of fun types.
 % Add support for them all eventually
+compat_ty({type, _, 'fun', [{type, _, any}, Res1]},
+          {type, _, 'fun', [{type, _, any}, Res2]},
+          A, TEnv) ->
+    compat(Res1, Res2, A, TEnv);
+compat_ty({type, _, 'fun', [{type, _, product, _Args1}, Res1]},
+          {type, _, 'fun', [{type, _, any}, Res2]},
```
[the remark this answers is not in the document]
Comment (gomoripeti) [reply src/typechecker.erl]: oh, indeed, I'll fix this (I did not want to move checking product/any in guard, but that is not even necessary)

File: src/typechecker.erl (hunk 1 above)
[the remark this answers is not in the document]
Comment (zuiderkwast) [reply src/typechecker.erl]: You're right.

**Comment by UlfNorell** [comment]
Can't look in detail right now, but

> This whole fun subtyping looks similar to subtyping of maps.

Note that function types are contravariant in the arguments and covariant in the result and maps are covariant in both keys and values. I.e.
```
        A₂ <: A₁   B₁ <: B₂
---------------------------------
fun((A₁) -> B₁) <: fun((A₂) -> B₂
```
whereas
```
    A₁ <: A₂   B₁ <: B₂
--------------------------
#{A₁ := B₁} <: #{A₂ := B₂}
```

**Comment by gomoripeti** [comment]
thanks, @UlfNorell , for clarifying. Although through examples I see the difference, but if I think like this:
Maps also assign values to keys, like functions assign return values to arguments. Does the difference comes from the meaning that `#{A := B}` means for ANY key of type A, a value of type B is assigned, while `fun( (A) -> B )` means this fun object should assign a value of type B to ALL possible args of type A?
somewhat connected: let's say I have a function `f` which returns an integer for all floats between 0..1. The spec could be`-spec f(float()) -> integer()`. Does this spec only tell me that if the arg is of type float the return value is guaranteed to be integer, or does it also tell me that `f` will handle all floats (which is not true - but there is no subtype of float in erlang with which I could express more precisely the domain)? And is there a difference between meaning of function specs and fun object types (aka lambda types or arrow types) in this aspect?

**Comment by UlfNorell** [comment]
The key difference between `fun((A) -> B)` and `#{ A := B }` is who supplies the values of type `A`. The map is essentially a list of `A`/`B` pairs, so all the `A`s are given by the map and accessible by the user. The function, on the other hand, is a promise to deliver a `B` given an `A` provided by the user.

In the map case, treating `#{ integer() := atom() }` as `#{ number() := atom() }` is perfectly fine. The worst thing that happens is that you look at the keys and find no floats. If you do the same with a function `fun((integer()) -> atom())` and start calling it with floats, it can crash and *justifiably blame you*.

Conversely, if you have a `#{ number() := atom() }` map and treat it as `#{ integer() := atom() }` you can get the list of keys as `[integer()]`, but they might all be floats. Treating a function from `number()` as a function from `integer()` is perfectly sensible and has no ill effects.

**Comment by zuiderkwast** [comment]
It's interesting, but contravariance for fun args is already implemented long ago. I think what's new here (apart from some cases with `fun()` and `fun((...) -> T)`) is how to handle subtyping when intersection types are involved. Currently, we don't.

E.g. how do we check that

```
fun( (integer()) -> integer();
     (float())   -> float() ) <: fun( (number()) -> number() )
```

**Comment by gomoripeti** [comment]
thanks @UlfNorell , now it's all clear. This is not something new, but I've been missing this fundamental bit.

For discussion about subtyping intersection types I opened #130

**Comment by gomoripeti** [comment]
there is an infinite loop in self-gradualizing with this change (I think somewhere in normalization) so this is not yet ready for merging

**Comment by zuiderkwast** [comment]
Beautiful commit 7349d8e!

**Comment by gomoripeti** [comment]
the "infinite loop" is now solved by #135 so this PR is ready to be merged.

**Pull request merged**